### ISSUE TITLE
Greatly unbalanced dataset. How to deal with it in chemprop?

### ISSUE THREAD

**Issue by ismorphism** [chemprop/chemprop#240]
Hello, everyone! I have a dataset, which is HUGELY unbalanced. Around 97% of compounds are non-active, and only 3% are active. Chemprop is overfitted to the data greatly. Do you have any advices how to deal with it? f1 metric is terrible...

**Comment by cjmcgill** [comment]
Hi @ismorphism. We recently added a loss function that might be helpful `--loss_function mcc`, which is based off of the Matthews Correlation Coefficient (related to F1). I think for your case you will also want to use the argument `--class_balance` and maybe increase `--batch_size <int: def=50>` so that you don't have training minibatches totally lacking in positive examples. We added mcc loss with this situation in mind and are hopeful that it may produce good results for you. We have also recently added f1 and mcc as evaluation metrics for when to carry out early stopping, which may be of interest to you.

Some other thoughts that are not as directly related to your problem but may still be relevant. Chemprop uses dropout for regularization and if you are experiencing a lot of overfitting, turning up the dropout in increments of 0.1 may be helpful `--dropout <float: def=0.0>`. Also if you are not using `--aggregation norm` or an ensemble model of 5 or so `--ensemble_size 5`, I would try both of those from a general model quality standpoint.

**Comment by ismorphism** [comment]
@cjmcgill thanks a lot for the reply! What do you think about `--split_type `? Could it affect the overfitting?

**Comment by cjmcgill** [comment]
When you tried mcc loss, did you also try `--class_balance`? That helps mcc loss be more stable. Same with the larger batch size. The thing it needs to avoid is ever having a mini batch in training that has zero positives in it (or zero negatives), because if that happens then it Nans. With `--class_balance`, you should be able to avoid that particular failure mode if your (number-of-positives) > (total-number)/(batch-size). Even better if you can make sure there's more than just one per batch. There are other ways that mcc can become numerically unstable too, but I think those two changes will help make those other failures less likely. 

I am not sure if there is a split type that would help overfitting.

**Comment by ismorphism** [comment]
@cjmcgill it definitely helps! Although I did not get `--aggregation norm` in the context of overfitting - how it could help? Increase in the value causes less overfitting?

**Comment by cjmcgill** [comment]
The aggregation norm isn't specifically about over fitting. It's just a good chance to get a better model overall, irrespective of the over fitting.

**Comment by ismorphism** [comment]
The higher the better - am I right?

**Comment by cjmcgill** [comment]
An additional couple of tips that a team member had for your problem. It may be beneficial to you to emphasize the positive values in different ways. One blunt but simple way would be manually removing some of your negative datapoints to rebalance your set. A bit more complicated would be using the data weighting option, where you can give extra weight to particular datapoints during training to emphasize the positives (see the data weighting and task weighting section in readme).

**Comment by ismorphism** [comment]
@cjmcgill Thanks for the advices, but data_weights and class_weights are useless unfortunately, they make F1 metrics worse

**Comment by cjmcgill** [comment]
I'm surprised that higher data_weights for the positives made the problem worse.

Another thought, did you use f1 or mcc as your metric while doing hyperparameter optimization? F1 as an available option for that is fairly recent addition to chemprop. And that might help you pick out better model parameters. The prc-auc should also be a metric that is less sensitive to imbalance so would be a good metric. You hyperparameter optimization might be a bit better if using that instead of binary_cross_entropy.

In general model improvement options, you could also look at using extra features. Like the rdkit 2d normalized feature generator which has been good for some people in the past. Or add your own features if you think you have something relevant.

On the whole, I think my best advice remains: use `--class_balance` with a suitably large batch size. And maybe try that in concert with mcc as the loss function.
